### ISSUE TITLE
Generic implementation of SVDQuant

### ISSUE THREAD

**Issue by dxqb** [nunchaku-tech/nunchaku#748]
I have implemented your core innovation of SVDQuant here in a way that might be more accessible for people, for use cases that can't benefit from your kernel fusion implementation:

https://github.com/Nerogar/OneTrainer/blob/1e8fecedddfd05ea33751d4f56e0eeef75372691/modules/module/quantized/LinearSVD.py

I cannot confirm the statement in your paper that it comes with a substantial performance impact. At least not for my usecase - about 5%.

Feel free to close this issue when you've seen it - unless you want to interpret it as a feature request to include a more accessible SVDQuant implementation in your repo.

**Comment by kabachuha** [comment]
@dxqb Is it possible to quantize Wan out of the box this way?

**Comment by dxqb** [comment]
> [@dxqb](https://github.com/dxqb) Is it possible to quantize Wan out of the box this way?

this is a developer tool not for end users, but it could be applied to any model. the point is to be able to use SVDQuant without having to combine it with other features of nunchaku like kernel fusion, which are quite complicated and not applicable to all use cases (for example LoRA training).

But this also means you cannot expect the same speed-ups as a full nunchaku model.

**Comment by dxqb** [comment]
without the kernel fusion and other nunchaku optimizations, you can achieve some significant speed up, but not near full nunchaku:

<img width="483" height="340" alt="Image" src="https://github.com/user-attachments/assets/1b3195ce-fea9-4bd9-a815-8b391c415c77" />

the advantage is that it doesn't require implementing each model but once implemented works on all models

**Comment by lmxyy** [comment]
@dxqb, How did you derive the 5% latency overhead? Is it measured against our W4A4 GEMM or the W16A16 baseline?

**Comment by dxqb** [comment]
> [@dxqb](https://github.com/dxqb), How did you derive the 5% latency overhead? Is it measured against our W4A4 GEMM or the W16A16 baseline?

It is measured against a W8A8 quant that uses 8-bit ops such as torch._int_mm / torch._scaled_mm, but not your fused kernels.
I can provide other measurements if you need it, but it is not surprising to me that the overhead is small, because the overhead of SVD Quant is identical to applying a LoRA that is not merged into the weights - which is also small.

This line is the overhead:
`return ((x @ self.svd_down.T) @ self.svd_up.T).to(x.dtype) + super().forward(x)`

How did you make your measurement? Did you use torch.compile? That could be the difference because I do, and torch.compile is quite good at fusing operations like that.

**Comment by dxqb** [comment]
As a side note, I wonder if you would agree with this. Am I missing something?

---

**Why no int4 / fp4?**
Nunchaku (https://github.com/nunchaku-tech/nunchaku) have shown that 4-bit quants are possible at no noticable quality loss. Even though their matrix math is faster than 8-bit quants, my benchmarks show that there is no performance gain overall for training just by employing 4-bit matrix math. The 8-bit linear math is already fast enough - most of the remaining time in a training step is spent on bf16 attention and other layers.
I believe the additional improved performance achieved by Nunchaku is due to their manual kernel fusion, not due to 4bit matmuls - but happy to be proven wrong.

**Comment by lmxyy** [comment]
> > [@dxqb](https://github.com/dxqb), How did you derive the 5% latency overhead? Is it measured against our W4A4 GEMM or the W16A16 baseline?
> 
> It is measured against a W8A8 quant that uses 8-bit ops such as torch._int_mm / torch._scaled_mm, but not your fused kernels. I can provide other measurements if you need it, but it is not surprising to me that the overhead is small, because the overhead of SVD Quant is identical to applying a LoRA that is not merged into the weights - which is also small.
> 
> This line is the overhead: `return ((x @ self.svd_down.T) @ self.svd_up.T).to(x.dtype) + super().forward(x)`
> 
> How did you make your measurement? Did you use torch.compile? That could be the difference because I do, and torch.compile is quite good at fusing operations like that.

And which device?

**Comment by lmxyy** [comment]
> > > [@dxqb](https://github.com/dxqb), How did you derive the 5% latency overhead? Is it measured against our W4A4 GEMM or the W16A16 baseline?
> > 
> > 
> > It is measured against a W8A8 quant that uses 8-bit ops such as torch._int_mm / torch._scaled_mm, but not your fused kernels. I can provide other measurements if you need it, but it is not surprising to me that the overhead is small, because the overhead of SVD Quant is identical to applying a LoRA that is not merged into the weights - which is also small.
> > This line is the overhead: `return ((x @ self.svd_down.T) @ self.svd_up.T).to(x.dtype) + super().forward(x)`
> > How did you make your measurement? Did you use torch.compile? That could be the difference because I do, and torch.compile is quite good at fusing operations like that.
> 
> And which device?

I see. 4070.

**Comment by lmxyy** [comment]
@sxtyzhangzk

**Comment by sxtyzhangzk** [comment]
When we measured the overhead of the LoRA branch in our SVDQuant paper, we compared the performance of a **single W4A4 linear layer**. If applied to a W8A8 layer, it is expected that the overhead is lower since the W8A8 main branch is slower than W4A4. 

Also, it is not surprising that the **end-to-end** overhead can be even lower since there are many operations other than linear layers in the network, especially if the benchmark is done using a generic implementation, in which case the rope/norm/activation functions take a significant part of the end-to-end latency. When we push everything to its limit and all other overheads are eliminated, the necessity of lora branch fusion will arise.

Anyway, it is good to see a generic implementation of SVDQuant that can benefit many other models. @lmxyy I think currently we have a `SVDQW4A4Linear`. Did we provide any method to directly quantize and convert a single linear layer into a W4A4 linear layer so that it can also be applied to other models?

**Comment by dxqb** [comment]
> When we measured the overhead of the LoRA branch in our SVDQuant paper, we compared the performance of a **single W4A4 linear layer**. If applied to a W8A8 layer, it is expected that the overhead is lower since the W8A8 main branch is slower than W4A4.
> 
> Also, it is not surprising that the **end-to-end** overhead can be even lower since there are many operations other than linear layers in the network, especially if the benchmark is done using a generic implementation, in which case the rope/norm/activation functions take a significant part of the end-to-end latency. When we push everything to its limit and all other overheads are eliminated, the necessity of lora branch fusion will arise.

:thumbsup: 

> 
> Anyway, it is good to see a generic implementation of SVDQuant that can benefit many other models. [@lmxyy](https://github.com/lmxyy) I think currently we have a `SVDQW4A4Linear`. Did we provide any method to directly quantize and convert a single linear layer into a W4A4 linear layer so that it can also be applied to other models?

that would be really nice to have, because the weights packing and activation calibration for W4A4 are not easy to understand now, and people might also want to use W4A4 without having to change model architecture and use kernel fusion.

**Comment by lmxyy** [comment]
I can take a look at this.

**Comment by zwukong** [comment]
Sorry to interrupt your discussion,but it is really excited to see your work on the generic implementation,which means more models will be smaller and faster

**Comment by dxqb** [comment]
Here is a generic implementation of SmoothQuant, that can be combined with SVDQuant as you do for nunchaku. I ended up not using it, because it doesn't seem to benefit LoRA training with W8A8 quants. Sometimes it even seems to hurt it. I don't know why, maybe there is a bug.
Anyway, in case this is useful to anyone:

[LinearSmooth.py](https://github.com/user-attachments/files/22863438/LinearSmooth.py)

**Comment by aslyansky-m** [comment]
@dxqb 
Your sample implementation might be very useful, do you have a complete calibration example for a full model?

**Comment by dxqb** [comment]
> [@dxqb](https://github.com/dxqb) Your sample implementation might be very useful, do you have a complete calibration example for a full model?

these are dev tools. if you are looking for a nunchaku model for your favourite model, this thread is not for you.
